akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/Buffer.scala

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.query.journal.leveldb
+
+import akka.annotation.InternalApi
+import akka.stream.Outlet
+import akka.stream.stage.GraphStageLogic
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[leveldb] trait Buffer[T] { self: GraphStageLogic =>
+  private var buf: Vector[T] = Vector.empty[T]
+  def buffer(element: T): Unit = {
+    buf :+= element
+  }
+  def buffer(all: Set[T]): Unit = {
+    buf ++= all
+  }
+  def deliverBuf(out: Outlet[T]): Unit = {
+    if (buf.nonEmpty && isAvailable(out)) {
+      val next = buf.head
+      push(out, next)
+      buf = buf.tail
```

Comment:
not sure `tail` for `Vector` is efficient, maybe use plain `java.util.LinkedList`?

Reply:
Ah ofc, previous one did bulk adds to the buffer. The main stages either add to the front or take the tail so maybe a scala linked list will do

Reply:
Or not they add to the end, will use a mutable list